Gems/Atom/RHI/Code/Include/Atom/RHI/MultiDeviceShaderResourceGroupData.h

```diff
@@ -0,0 +1,577 @@
+
+            return m_deviceShaderResourceGroupDatas.at(deviceIndex);
+        }
+
+        //! Reset image and buffer views setup for this MultiDeviceShaderResourceGroupData
+        //! So it won't hold references for any RHI resources
+        void ResetViews();
+
+        //! Returns the shader resource layout for this group.
+        const ShaderResourceGroupLayout* GetLayout() const;
+
+        using ResourceType = ShaderResourceGroupData::ResourceType;
+
+        using ResourceTypeMask = ShaderResourceGroupData::ResourceTypeMask;
+
+        //! Reset the update mask
+        void ResetUpdateMask();
+
+        //! Enable compilation for a resourceType specified by resourceTypeMask
+        void EnableResourceTypeCompilation(ResourceTypeMask resourceTypeMask);
+
+        //! Returns the mask that is suppose to indicate which resource type was updated
+        uint32_t GetUpdateMask() const;
+
+        //! Update the indirect buffer view with the indices of all the image views which reside in the global gpu heap.
+        void SetBindlessViews(
+            ShaderInputBufferIndex indirectResourceBufferIndex,
+            const MultiDeviceBufferView* indirectResourceBufferView,
+            AZStd::span<const MultiDeviceImageView* const> imageViews,
+            uint32_t* outIndices,
+            AZStd::span<bool> isViewReadOnly,
+            uint32_t arrayIndex = 0);
+
+        //! Update the indirect buffer view with the indices of all the buffer views which reside in the global gpu heap.
+        void SetBindlessViews(
+            ShaderInputBufferIndex indirectResourceBufferIndex,
+            const MultiDeviceBufferView* indirectResourceBufferView,
+            AZStd::span<const MultiDeviceBufferView* const> bufferViews,
+            uint32_t* outIndices,
+            AZStd::span<bool> isViewReadOnly,
+            uint32_t arrayIndex = 0);
+
+    private:
+        bool ValidateSetImageView(ShaderInputImageIndex inputIndex, const MultiDeviceImageView* imageView, uint32_t arrayIndex) const;
+        bool ValidateSetBufferView(ShaderInputBufferIndex inputIndex, const MultiDeviceBufferView* bufferView, uint32_t arrayIndex) const;
+
+        template<typename TShaderInput, typename TShaderInputDescriptor>
+        bool ValidateImageViewAccess(TShaderInput inputIndex, const MultiDeviceImageView* imageView, uint32_t arrayIndex) const;
+        template<typename TShaderInput, typename TShaderInputDescriptor>
+        bool ValidateBufferViewAccess(TShaderInput inputIndex, const MultiDeviceBufferView* bufferView, uint32_t arrayIndex) const;
+
+        //! Device mask denoting on which devices the SRG data is needed
+        MultiDevice::DeviceMask m_deviceMask;
+
+        ConstPtr<ShaderResourceGroupLayout> m_shaderResourceGroupLayout;
+
+        //! Mask used to check whether to compile a specific resource type. This mask is managed by RPI and copied over to the RHI every
+        //! frame.
+        uint32_t m_updateMask = 0;
```

Comment:
Why does this class needs its own version of m_shaderResourceGroupLayout and m_updateMask? It should just be forwarding the calls to the SD version of ShaderResourceGroupData. Maybe the validation code can be here and removed from the SD version when we transition over.

Reply:
The reason for that is because we need it for the respective getters and methods that are accessing these variables.

We could of course forward those to one or multiple of the SD objects. However, the question then is either to which one or how do you aggregate the result if it differs (though it shouldn't, but nothing is preventing someone from getting a SD object and changing something just there). Also what should happen if the device mask and the available devices result in no SD object being available at all, i.e., the map being empty?

Therefore, we decided to duplicate those variables and their handling in order to have well defined behavior in all cases, even if no SD objects are actually created or if specific SD objects differ in those variables.

However, if you think we should do otherwise, we are open to suggestions.

Reply:
Eventually, I think the MD version of SRGData should have api to forward the data update to a specific SD version. But for now in order to keep this class to be minimal and not have duplicated code unless we have a use case for it I propose we just update all the SD versions equally. So the assumption is that when the higher level code updates a MDSRgData all SD versions get updated equally. We should be able to test MGPU support just based off this assumption. Eventually the we want to build different forms of MGPU pipeline we can update the API with a mask in order to tell it which SD versions to update.

Reply:
Agreed! Is there anything else needed for approval?

Reply:
That should be good enough. Let's remove all the duplicated code related to it and I think the PR should be good for approval.

Reply:
I was able to remove `m_updateMask`, as all calls can be properly forwarded to the SD versions, so that is gone now. 
`m_shaderResourceGroupLayout` is a different story though, as `GetLayout() ` is called frequently and with no clear way of knowing, which SD version should return its corresponding layout.
Similarly, for the `FindShaderInput*` we would have to choose from which SDSRGData we would want to retrieve the index.
We would suggest keeping it as it is for the moment and changing it later if needed.

Reply:
ok. This is fine for now.